### ISSUE TITLE
Browse Alphabetically is slow

### ISSUE THREAD

**Issue by nmswanson** [CARLI/vufind#348]
ISU - Browse Alphabetically is slow

email 4/19/18 page 13

[on 6/25/18 ISU local browse alphabetically searches were usually around 10-15 seconds, some were longer ex. Tchaikovsky (about 29 sec.) or much shorter (4 sec.)]

**Comment by cedelis** [comment]
Yes, it seems that the further along you page through results, the slower the Alphabrowse response becomes. The algorithm in the stock browse-handler code does have room for improvement: for each "Next" page click, the browse-handler executes the original search query (e.g., Tchaikovsky) each time and then skips the required offset number (which depends on the page number) of records from the original search result. 

One idea that might work (which will require some further research and testing, of course): Split the algorithm up into two separate types of searches:

1) Query search: if there is a search term present (e.g., from=Tchaikovsky), this means that a search is being initiated; therefore, perform the query as usual (original algorithm); 

2) Direct index retrieval: if there is no search term present (i.e., the patron has either clicked on "Next" or "Prev"), then there will instead need to exist a new parameter, say "searchOffset," which is the index offset of the search query's initial result. This would allow the handler to bypass the search altogether and go directly to the correct result each time (remember: the Alphabrowse database is already sorted, i.e., "left-anchored"; so let's take advantage of this fact).

In order to implement this strategy, the "Next" link (e.g., for page 3) would need to change from:

offset=40&from=Tchaikovsky&source=author&rows=20

to:

searchOffset=9932687&offset=40&source=author&rows=20

Note that: 

a) the "&from=Tchaikovsky" parameter is missing

b) the value 9932687 comes from the first matching result of the "&from=Tchaikovsky" query 

Here's where that number (9932687) comes from (backend Solr query):

E.g., curl -q 'http://server:port/solr/biblio/browse?prependFromField=institution&json.nl=arrarr&offset=40&prependFromValue=ISUdb&from=tchaikovsky&offset=40&source=author&rows=20&wt=xml'


```
<?xml version="1.0" encoding="UTF-8"?>
<response>
  <lst name="responseHeader">
    <int name="status">0</int>
    <int name="QTime">63</int>
  </lst>
  <lst name="Browse">
    <int name="startRow">9932687</int>
    <int name="offset">40</int>
    <str name="matchType">NONE</str>
    <int name="totalCount">108017</int>
...
  </lst>

```
